### ISSUE TITLE
A possibility for the surveyors to download their own data

### ISSUE THREAD

**Issue by ekol-ali** [biodiversitydata-se/biocollect#199]
Several surveyors have asked for the possibility to download their own data, as text or Excel-file. As Admins we agree totally with this wish. This is of **no priority now**, but further on when the basics have settled, it is definitely something that we and the surveyors want.

There is already such a function for route coordinates, e.g.: 
https://biocollect.biodiversitydata.se/sft/site/index/414a40c8-0ffc-44ab-8c9d-a79959b80d33

As far as their bird data are concerned, there should probably be one downlaod option per scheme (vinterpunktrutter, standardrutter, etc) since they all have different no. of coloumns etc. When you ask for vinterpunktrutter, you get all your surveys of vinterpunktrutter.

**Comment by mathieuLU** [comment]
Answers from ALA : 
(Chris)
`The smallest unit MERIT (and I think BioCollect) can export is activity data of the same type for a single project.

So the code to produce the export is there, you’ll just need to implement the code that specifies the criteria for the export.

Sounds like it would be a useful feature to me.`

(Temi)
` 
It should be possible to do that as long as the download URL is constructed properly. You should be able to get examples of download query by going to Project’s data tab and clicking the download button. Append a filter query with activity id should do the trick.`


Use this example from ALA
https://biocollect.ala.org.au/acsa/project/index/06b823b6-f259-4e80-9441-84e7c68cc09d
A click on the download button calls => 
https://biocollect.ala.org.au/bioActivity/downloadProjectData?projectId=06b823b6-f259-4e80-9441-84e7c68cc09d&hub=acsa&max=10&offset=0&sort=lastUpdated&order=desc&flimit=15&view=project&spotterId=&projectActivityId=&clientTimezone=Europe%2FStockholm&searchTerm=



I tried for STD : 
https://biocollect.biodiversitydata.se/bioActivity/downloadProjectData?projectId=89383d0f-9735-4fe7-8eb4-8b2e9e9b7b5c&hub=sft&max=10&offset=0&sort=lastUpdated&order=desc&flimit=15&view=project&spotterId=&projectActivityId=&clientTimezone=Europe%2FStockholm&searchTerm=

It does not work but I will investigate

**Comment by mathieuLU** [comment]
A discussion is needed here.

Which format do you want ?
my first thought was that you wanted for a user to have a download button on a survey page.
But Fredrik had in mind something like a "recap button" that could sum up all the data recorded by a user for al the routes of a specific program.

Layout and output to be discussed @ekol-ali

**Comment by ekol-ali** [comment]
I suggest, after our discusison, that you start by creating an Excel-file where the person gets ALL DATA: 1 file, with 1 spreadsheet per survey type. Use the template you use in the export-function (SQL) on Intranet, but of course with a minimum of "internal" stuff. We can discuss which columns should be included.

**Comment by mathieuLU** [comment]
We discussed this with Fredrik this morning as well. 
If you could create an example file with what you want to have in it for a specific user, that would be great.

Then we can design a page on BioCollect from which user can : 
 - specify which delprogram they want to get data from (checkboxes ?)
 - request the excel file

Since it won't be an existing BioCollect page converted to an excel file, but a custom extract, the script should be executed from the intranet side. So the user requests from BC, and then once or twice a day the intranet generates the requested files and send them to the email addresses. This is a suggestion to be discussed ;-)

**Comment by mathieuLU** [comment]
Recap of the meeting today : 

 - I have the 2 files you created with the expected template for each delprogram. 
 - i'll make sure to keep the same vocabulary in the files than in BioCollect. As detailled in Annelie's mail
 - i'll add some columns to get links to BioCollect (link to the protokoll, link to the lokal, ...)
 - only the "approved" data
 - no history of the changes made on the surveys
 - we'll have to write a "help" tab in the excel template. With the explanations of the rows 000

For a surveyor, the process to get the data : 
homepage => "Se vad jag rapporterat tidigare " => download button on the top right panel 
![Screenshot from 2024-10-02 14-38-43](https://github.com/user-attachments/assets/165ae0c5-7275-464a-b382-a275a7f863a1)

IF POSSIBLE, the download should use the filters on the left panel, to only download what is selected here.
Maybe the button could adapt to the numbers of surveys available "Download XX surveys in excel format"

The same template can be used for downloading one specific survey.
https://biocollect.biodiversitydata.se/sft/bioActivity/index/280b9d11-5664-4a3b-ac26-3747be151054?returnTo=%2Fsft%2FbioActivity%2Flist
with a download button like that : 
![Screenshot from 2024-10-02 14-42-32](https://github.com/user-attachments/assets/ec940d3c-eb10-4996-bfad-eaf85f806a00)

or directly from the survey list for one simple download ?
![Screenshot from 2024-10-02 14-42-50](https://github.com/user-attachments/assets/71a41c0c-62b4-4e5e-b195-bcbf5bf55ce4)

**Comment by ekol-ali** [comment]
Hej Mathieu

en god sammanfattning!

Kör på!

Åke

Från: mathieuLU ***@***.***>
Skickat: Wednesday, 2 October 2024 14:46
Till: biodiversitydata-se/biocollect ***@***.***>
Kopia: Åke Lindström ***@***.***>; Mention ***@***.***>
Ämne: Re: [biodiversitydata-se/biocollect] A possibility for the surveyors to download their own data (Issue #199)


Recap of the meeting today :

  *   I have the 2 files you created with the expected template for each delprogram.
  *   i'll make sure to keep the same vocabulary in the files than in BioCollect. As detailled in Annelie's mail
  *   i'll add some columns to get links to BioCollect (link to the protokoll, link to the lokal, ...)
  *   only the "approved" data
  *   no history of the changes made on the surveys

For a surveyor, the process to get the data :
homepage => "Se vad jag rapporterat tidigare " => download button on the top right panel
Screenshot.from.2024-10-02.14-38-43.png (view on web)<https://github.com/user-attachments/assets/165ae0c5-7275-464a-b382-a275a7f863a1>

IF POSSIBLE, the download should use the filters on the left panel, to only download what is selected here.
Maybe the button could adapt to the numbers of surveys available "Download XX surveys in excel format"

The same template can be used for downloading one specific survey.
https://biocollect.biodiversitydata.se/sft/bioActivity/index/280b9d11-5664-4a3b-ac26-3747be151054?returnTo=%2Fsft%2FbioActivity%2Flist
with a download button like that :
Screenshot.from.2024-10-02.14-42-32.png (view on web)<https://github.com/user-attachments/assets/ec940d3c-eb10-4996-bfad-eaf85f806a00>

or directly from the survey list for one simple download ?
Screenshot.from.2024-10-02.14-42-50.png (view on web)<https://github.com/user-attachments/assets/71a41c0c-62b4-4e5e-b195-bcbf5bf55ce4>

**Comment by mathieuLU** [comment]
Make sure that it's possible as well for the medinventerare

**Comment by mathieuLU** [comment]
ok so a first version runs locally now !
 - only for standardrutter
 - using some filters from the left panel (site names, years)
 - returning an excel file according to the template you sent me some months ago

i'll try to put something on production tomorrow, visible only for admins

**Comment by mathieuLU** [comment]
recap of the past 2 weeks : 
 - now possible to get data from 5 of the 6 projects (STD/KUST/VINTER/SOMMAR/IWC). Only missing NATT now
 - a limit is set to 5000 documents to avoid the site to crash (maybe better to add a limit on the "allrecords" page as well ?)
 - statistics module to store the requests + params + date + number of docs
 - filename with userId + date
 - configuration file containing the variable
 - protection in case of missing strange to avoid the code to crash
 - download implemented from the page "Läsa om de olika delprogrammen"

remain : 
 - publish it for the users when ok

feedbacks appreciated :)

**Comment by mathieuLU** [comment]
One important thing about the filer : so far the filters implemented are  
 - delprogram (array)
 - år (array)
 - lokal (array)
 - personId (for "myrecords")

so some are not implemented. Are they needed ?
like : filter by month, filter by species

**Comment by ekol-ali** [comment]
I do not think we need filters on Month and Species. Any surveyor who wants to do exactly that can manage that in a second step. 
If it is easy to implement - YES, if a lot of work - NO.

**Comment by mathieuLU** [comment]
so it's all published.
In the last commit i cleaned some stuff to make sure the token to the module is in the config file.
https://github.com/biodiversitydata-se/biocollect/commit/44ce6cc6c7143f6c470ab2e35a15915d81d38a96

some users are testing right now

all good on my side, i think we can close the issue if everyone is happy :)
